tabular/src/autogluon/tabular/models/tab_transformer/TabTransformer.py

```diff
@@ -0,0 +1,160 @@
+                p_dropout=self.p_dropout,
+                minimize_emb_dim=False,
+                drop_whole_embeddings=self.drop_whole_embeddings,
+                one_hot=False,
+                out_dim=self.hidden_dim,
+                shared_embedding=column_embedding,
+                n_shared_embs=n_shared_embs,
+                shared_embedding_added=shared_embedding_added,
+            )
+        if self.n_cont_features:
+            self.cont_norm = self.get_norm(self.n_cont_features)
+            self.cont_initializer = nn.Linear(self.n_cont_features, hidden_dim * n_cont_embeddings)
+            self.cont_init_norm = self.get_norm(hidden_dim * n_cont_embeddings)
+
+        if self.readout == 'readout_emb':
+            self.readout_emb = nn.Parameter(
+                torch.zeros(1, hidden_dim).uniform_(-1, 1))  # We do the readout from a learned embedding
+            self.n_embeddings += 1
+
+        if fix_attention is True:
+            self.n_cat_embeddings = len(self.cat_feat_origin_cards)
+            self.tfmr_layers = nn.ModuleList([TransformerClass.TransformerEncoderLayer_modified(d_model=hidden_dim,
+                                                                               n_cat_embeddings=self.n_cat_embeddings,
+                                                                               nhead=n_heads,
+                                                                               dim_feedforward=4 * hidden_dim,
+                                                                               dropout=self.p_dropout,
+                                                                               activation='gelu') for _ in
+                                              range(n_layers)])
+        else:
+            self.tfmr_layers = nn.ModuleList([nn.TransformerEncoderLayer(d_model=hidden_dim,
+                                                                         nhead=n_heads,
+                                                                         dim_feedforward=4 * hidden_dim,
+                                                                         dropout=self.p_dropout,
+                                                                         activation='gelu') for _ in
+                                              range(n_layers)])
+   
+
+    def init_input(self, input):
+        try_import_torch()
+        import torch
+        feats = [init(input[:, i]) for i, init in enumerate(self.cat_initializers.values())]
+
+        if self.readout == 'readout_emb':
+            readout_emb = self.readout_emb.expand_as(feats[0])
+            feat_embs = torch.stack([readout_emb] + feats,
+                                    dim=0)  # (n_feat_embeddings + 1) x batch x hidden_dim
+        else:
+            feat_embs = torch.stack(feats, dim=0)  # n_feat_embeddings x batch x hidden_dim
+        return feat_embs
+
+    def run_tfmr(self, feat_embs):
+        try_import_torch()
```

Comment:
general comment: I don't think all of these checks are necessary. Just have one check in the model which extends AbstractModel during .fit, all this code should only be imported after that one check has been done.

Reply:
Compressed all the lazy imports down to one in TabTransformerModel class.